src/app/child-dev-project/reporting/query.service.ts

```diff
@@ -0,0 +1,213 @@
+    });
+  }
+
+  /**
+   * Returns all entities which reference a entity from the passed list of entities (by their id)
+   * @param srcEntities the entities for which relations should be found
+   * @param entityType the type of entities where relations should be looked for
+   * @param relationKey the name of the attribute that holds the reference.
+   *                    The attribute can be a string or a list of strings
+   * @returns a list of the related unique entities
+   */
+  private getRelated(
+    srcEntities: Entity[],
+    entityType: string,
+    relationKey: string
+  ): Entity[] {
+    const targetEntities = this.toArray(this.entities[entityType]);
+    const srcIds = srcEntities.map((entity) => entity.getId());
+    if (
+      targetEntities.length > 0 &&
+      Array.isArray(targetEntities[0][relationKey])
+    ) {
+      return targetEntities.filter((entity) =>
+        (entity[relationKey] as Array<string>).some((id) =>
+          srcIds.includes(id.split(":").pop())
+        )
+      );
+    } else {
+      return targetEntities.filter((entity) =>
+        entity[relationKey]
+          ? srcIds.includes(entity[relationKey].split(":").pop())
+          : false
+      );
+    }
+  }
+
+  private filterByObjectAttribute(
+    objs: any[],
+    attr: string,
+    key: string,
+    value: string
+  ): any[] {
+    const values = value.replace(new RegExp(" ", "g"), "").split("|");
+    return objs.filter((obj) => {
+      if (obj?.hasOwnProperty(attr)) {
+        return values.includes(obj[attr][key]?.toString());
+      }
+      return false;
+    });
+  }
+
+  private getIds(objs: any[], key: string): string[] {
+    const ids: string[] = [];
+    objs.forEach((obj) => {
+      if (obj.hasOwnProperty(key)) {
+        ids.push(...obj[key]);
+      }
+    });
+    return ids;
+  }
+
+  /**
+   * Return the ids of all the participants of the passed events with the defined attendance status using the `countAs`
+   * attribute. The list may contain duplicates and the id does not necessarily have the entity prefix.
+   * @param events the array of events
+   * @param attendanceStatus the status for which should be looked for
+   * @returns the ids of children which have the specified attendance in an event
+   */
+  private getParticipantsWithAttendance(
+    events: EventNote[],
+    attendanceStatus: string
+  ): string[] {
+    const attendedChildren: string[] = [];
+    events.forEach((e) =>
+      e.children.forEach((childId) => {
+        if (e.getAttendance(childId).status.countAs === attendanceStatus) {
+          attendedChildren.push(childId);
+        }
+      })
+    );
+    return attendedChildren;
+  }
+
+  /**
+   * Returns the active child school relations from the input array
+   * @param relations an array of child school relations
+   * @returns an array of active child school relations
+   */
+  private getActive(relations: ChildSchoolRelation[]): ChildSchoolRelation[] {
```

Comment:
can and should this also be used for Child.isActive? The current demo report actually includes inactive children, thereby showing a different number than the dashboard widget, for example.

Reply:
Hmm I am not sure if `isActive` makes sense for all classes (why would a note be active or not) but I see that in some cases its reasonable. Concerning the dashboard the two numbers refer to a different entity (active children based on the dropout and active child-school-relations based on the end date) therefore the numbers differ in the case where a child that didnt dropout has not visited any events in the recent period.